### PR TITLE
[Core] Task level config

### PR DESCRIPTION
**Pull request by Michaelvll**
Allow task-level configuration.

~~Blocked by #3682~~

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - [x] `pytest tests/test_config.py`
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

### CODE REVIEW

File: sky/resources.py [hunk 1]
```diff
@@ -26,6 +26,14 @@
 
 _DEFAULT_DISK_SIZE_GB = 256
 
+OVERRIDEABLE_CONFIG_KEYS = [
+    ('docker',),
```
Comment (cblmemo) [inline sky/resources.py:30]:
```suggestion
    ('docker', 'run_options'),
```

File: sky/skypilot_config.py
Comment (cblmemo) [inline sky/skypilot_config.py:1]: should we update the docstr for this module as well

File: sky/task.py [hunk 2]
```diff
@@ -456,8 +456,25 @@ def from_yaml_config(
...
+                         f'{cluster_config_override}')
+        assert not experimnetal_configs, ('Invalid task args: '
+                                          f'{experimnetal_configs.keys()}')
+
         # Parse resources field.
-        resources_config = config.pop('resources', None)
+        resources_config = config.pop('resources', {})
+        if cluster_config_override is not None:
+            assert resources_config.get('_cluster_config_overrides') is None, (
+                'Cannot set _cluster_config_overrides in both resources and '
```
Comment (cblmemo) [inline sky/task.py:474]: Should we raise an error if we find user specified `resources._cluster_config_overrides` as this is not allowed (according to the comment in the schema)?

File: sky/utils/schemas.py [hunk 3]
```diff
@@ -370,6 +375,70 @@ def get_service_schema():
     }
 
 
+def _filter_schema(schema: dict, keys_to_keep: dict) -> dict:
```
Comment (cblmemo) [inline sky/utils/schemas.py:378]:
```suggestion
def _filter_schema(schema: dict, keys_to_keep: List[Tuple[str, ...], ...]) -> dict:
```

File: sky/utils/schemas.py [hunk 4]
```diff
@@ -370,6 +375,70 @@ def get_service_schema():
...
+
+        for key, sub_schema in current_schema['properties'].items():
+            if key in current_path_dict:
+                # Recursively keep keys if further path dict exists
+                new_schema['properties'][key] = {}
+                new_schema['properties'][key] = keep_keys(
+                    sub_schema, current_path_dict[key],
+                    new_schema['properties'][key])
+
+        return new_schema
```
Comment (cblmemo) [inline sky/utils/schemas.py:420]: Seems like we do not handle the `required` and `oneOf` field? Should we only keep those in the path_dict?

**Review by cblmemo — COMMENTED** [review]
Thanks for adding this feature @Michaelvll ! Left some comments to discuss :)

File: sky/utils/schemas.py (hunk 4 above)
Comment (Michaelvll) [reply sky/utils/schemas.py]: Good point! Excluded those fields for now, as we don't need to support them.

File: sky/task.py (hunk 2 above)
Comment (cblmemo) [reply sky/task.py:474]: Good point! Lets keep the current one

**Review by cblmemo — APPROVED** [review]
Thanks for the fix @Michaelvll ! LGTM.

**Pull request merged**